### DISCUSSION TITLE
Spacy keep asking for sentencizer when training EntityLinker component

### DISCUSSION THREAD

**Discussion by wy-yip** [explosion/spaCy#7855] [Help: Coding & Implementations] [answered]
I started using spaCy last week to train an EntityLinker component for standardizing shop names. When I was training the EntityLinker component, the model kept asking for sentencizer even though the parser component had defined sentences boundaries when I ran it separately. 

The code I used for training (sent_model is a model whose NER component I trained with my own labels):

```
# Disable pipeline components you dont need to change
pipe_exceptions = ["parser", "entity_linker"]
unaffected_pipes = [pipe for pipe in sent_model.pipe_names if pipe not in pipe_exceptions]

with sent_model.disable_pipes(*unaffected_pipes):   # train only the entity_linker and parser
    optimizer = sent_model.resume_training()
    for itn in range(30):   
        random.shuffle(TRAIN_DATA)
        batches = minibatch(TRAIN_DATA, size=compounding(4.0, 32.0, 1.001))  
        losses = {}
        for batch in batches:
            sent_model.update(
                batch,   
                drop=0.2,     
                losses=losses,
                sgd=optimizer,
            )
            print(itn, "Losses", losses)
```

The error: 
```
ValueError: [E030] Sentence boundaries unset. You can add the 'sentencizer' component to the pipeline with: `nlp.add_pipe('sentencizer')`. Alternatively, add the dependency parser or sentence recognizer, or set sentence boundaries by setting `doc[i].is_sent_start`.
```

I created my Example object using the code below. The NER label start and end are stored in a pandas frame, which I then use to match the token positions with reference to this post: https://github.com/explosion/spaCy/issues/4158 

```
TRAIN_DATA=[]

nlp = spacy.load("it_core_news_md")

docs = list(nlp.pipe(list(el_train_set['sentence'])))

for index,doc in enumerate(docs):
    
    chars_to_tokens = {}
    
    for token in doc:
        for i in range(token.idx, token.idx + len(token.text)):
            chars_to_tokens[i] = token.i
        
    token_start = chars_to_tokens[el_train_set.loc[index,'start']]
    token_end = chars_to_tokens[el_train_set.loc[index,'end']-1]
    span = Span(doc,token_start,token_end+1)
    
    ID = el_train_set.loc[index,'clean_names']
    
    annots = {
        'links': {(span.start_char,span.end_char):{ID:1.0}},
    }
    
    TRAIN_DATA.append(Example.from_dict(doc, annots))
```

I thought it was because my training data does not have sentence boundaries, but I do get sentences from the docs:
```
for sent in docs[0].sents:
    print(sent)
```

and I got 
```
facebk *ailassociazion:
Merchant type is STANDARD.
Mcc category code is 8398.
Country code is IRL.
Prefix is FACEBK.
```

I also ran an analysis on the pipeline components:
All the attributes required by entity_linker are present, especially doc.sents (assigned by parser) which spacy says is absent. 

```
============================= Pipeline Overview =============================

#   Component         Assigns               Requires         Scores             Retokenizes
-   ---------------   -------------------   --------------   ----------------   -----------
0   tok2vec           doc.tensor                                                False      
                                                                                           
1   morphologizer     token.morph                            pos_acc            False      
                      token.pos                              morph_acc                     
                                                             morph_per_feat                
                                                                                           
2   tagger            token.tag                              tag_acc            False      
                                                                                           
3   parser            token.dep                              dep_uas            False      
                      token.head                             dep_las                       
                      token.is_sent_start                    dep_las_per_type              
                      doc.sents                              sents_p                       
                                                             sents_r                       
                                                             sents_f                       
                                                                                           
4   ner               doc.ents                               ents_f             False      
                      token.ent_iob                          ents_p                        
                      token.ent_type                         ents_r                        
                                                             ents_per_type                 
                                                                                           
5   entity_linker     token.ent_kb_id       doc.ents         nel_micro_f        False      
                                            doc.sents        nel_micro_r                   
                                            token.ent_iob    nel_micro_p                   
                                            token.ent_type                                 
                                                                                           
6   attribute_ruler                                                             False      
                                                                                           
7   lemmatizer        token.lemma                            lemma_acc          False      

✔ No problems **found.**
```

Thank you for your time!

**Comment by svlandeg** [comment, marked as answer]
Hi!

Sorry that this has been frustrating. Conceptually, the `entity_linker` should indeed be able to use the sentences set by the `parser`.

However, the way that training currently works in spaCy 3.0, is that each component is updated separately from the training data. That means that during training, each component does not have access to predictions of another component. This is why you actually need to set gold-standard sentence boundaries on your training data for the `entity_linker`. The error message should be more specific about that. https://spacy.io/api/data-formats#dict-input shows example input data for the entity linker.

Now, all that said, we realise this situation is not ideal, and we've been working on a bit of a refactor of the training process: https://github.com/explosion/spaCy/pull/7767. When this PR gets merged, pipeline components will be able to set annotations during training, just as they would do at runtime. This will allow us to solve this more easily, you'll be able to rely on the predicted sentence boundaries from the parser instead of having to annotate them in your training data. This feature will be available from 3.1 onwards, and we'll revise the entity linker code accordingly.

For now, as a workaround, you can set sentences on the `Example.reference` doc object by running a sentencizer or parser on it, before feeding that Example into the training loop.

**Reply by wy-yip** [reply]
No worries - thank you for the amazing project. Would you mind providing some pseudo-code/ pointing me to a code example to annotate sentence boundaries if I am not using prodigy? Sorry am new both to programming and spaCy!

**Reply by adrianeboyd** [reply]
I'm confused since it looks like sentences are being added to the predicted examples in `TRAIN_DATA` from `it_core_news_md`.

Note that when training from a config you would have to write a custom corpus reader in v3.0 because `spacy.Corpus.v1` uses `nlp.make_doc()` (just tokenization) for the predicted docs.

**Reply by adrianeboyd** [reply]
Ah, wait, nevermind, it's the reference docs.

**Reply by svlandeg** [reply]
@wy-yip : following a similar code style as in your original post, you'd have something like this:
```
    ID = "Q2146908"
    TYPE = "PERSON"
    annots = {
        "links": {},
        "entities": []
    }
    for start_char, end_char in my_spans:
        annots["links"][(start_char, end_char)] = {ID: 1.0}
        annots["entities"].append((start_char, end_char, TYPE))
    ex = Example.from_dict(nlp.make_doc(text), annots)
```
This creates the `Example` from the annotation dict - which also needs to define the entity offsets with the keyword `entities`, as you can see in the code above.
Then, you can add sentence annotation to the `reference` doc in the `Example` like so:
```
    sentencizer = nlp.add_pipe("sentencizer")  # or grab the parser, or any other method...
    ex.reference = sentencizer(ex.reference)
```
and then you'd do as before:
```
TRAIN_DATA.append(ex)
```
Let us know if you run into issues!

**Reply by wy-yip** [reply]
Thank you for the helpful code snippets! 

I tried re-training the data with the following code:

```
TRAIN_DATA=[]

nlp = spacy.load("it_core_news_md")
sentencizer = nlp.add_pipe('sentencizer')

docs = list(nlp.pipe(list(el_train_set['sentence'])))


for index,doc in enumerate(docs):
    
    chars_to_tokens = {}
    
    for token in doc:
        for i in range(token.idx, token.idx + len(token.text)):
            chars_to_tokens[i] = token.i
        
    token_start = chars_to_tokens[el_train_set.loc[index,'start']]
    token_end = chars_to_tokens[el_train_set.loc[index,'end']-1]
    span = Span(doc,token_start,token_end+1)
    
    ID = el_train_set.loc[index,'clean_names']
    
    annots = {
        'links': {(span.start_char,span.end_char):{ID:1.0}},
        'entities': [(span.start_char, span.end_char, 'ORG')]
    }
    
    ex = Example.from_dict(doc, annots)
    ex.reference = sentencizer(ex.reference)
        
    TRAIN_DATA.append(ex)
    

    
```

which did annotate sentence boundaries as I skimmed at the training data (sent_start), but when I ran the training model again

```
pipe_exceptions = ["tok2vec", "ner", 'parser', "entity_linker"]
unaffected_pipes = [pipe for pipe in sent_model.pipe_names if pipe not in pipe_exceptions]

with sent_model.disable_pipes(*unaffected_pipes):   
    optimizer = sent_model.resume_training()
    for itn in range(30):   
        random.shuffle(TRAIN_DATA)
        batches = minibatch(TRAIN_DATA, size=compounding(4.0, 32.0, 1.001))  # increasing batch sizes
        losses = {}
        for batch in batches:
            sent_model.update(
                batch,   
                drop=0.5,      # prevent overfitting
                losses=losses,
                sgd=optimizer,
            )
            print(itn, "Losses", losses)
```

The ValueError becomes a blank 

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-22-18657989d108> in <module>
     14         losses = {}
     15         for batch in batches:
---> 16             sent_model.update(
     17                 batch,
     18                 drop=0.5,      # prevent overfitting

~\anaconda3\lib\site-packages\spacy\language.py in update(self, examples, _, drop, sgd, losses, component_cfg, exclude)
   1105             if name in exclude or not hasattr(proc, "update"):
   1106                 continue
-> 1107             proc.update(examples, sgd=None, losses=losses, **component_cfg[name])
   1108         if sgd not in (None, False):
   1109             for name, proc in self.pipeline:

~\anaconda3\lib\site-packages\spacy\pipeline\transition_parser.pyx in spacy.pipeline.transition_parser.Parser.update()

~\anaconda3\lib\site-packages\spacy\pipeline\transition_parser.pyx in spacy.pipeline.transition_parser.Parser.get_batch_loss()

~\anaconda3\lib\site-packages\spacy\pipeline\_parser_internals\ner.pyx in spacy.pipeline._parser_internals.ner.BiluoPushDown.set_costs()

ValueError: 
``` 

I experimented with different pipeline components disabled - whenever I disabled `ner` or `tok2vec`, the error `Parameter 'E' for model 'hashembed' has not been allocated yet.` came up. The ValueError remained blank regardless of whether the `parser` component is disabled or not.

**Reply by svlandeg** [reply]
This second error seems not to be directly connected to the `entity_linker` component. Do you run into the same issue when you remove the `entity_linker` from the pipeline? In fact - would it be possible for you to create a minimal code snippet that still exhibits this error, but that I can run on my end to reproduce? That will help with debugging. Right now, I'm not entirely sure exactly how you're defining your pipeline, so it would help to see everything.

Also, for spaCy v3, we do recommend to use the [training config system](https://spacy.io/usage/training#config) instead of writing your own training loops, exactly to prevent these kind of errors around initialization of internal weights etc. Once you get the hang of it, you'll notice that it's actually quite powerful & customizable.

**Reply by wy-yip** [reply]
Hmm I used one example to run the code on and the error is still there (to leave no stone unturned, I am printing the whole code chunk here, but I am suspecting there is something wrong with the way I labelled the training data):

```
import pandas as pd
import csv
from pathlib import Path
from collections import defaultdict # for working with dict in python
import spacy
from spacy.kb import KnowledgeBase
from spacy.tokens import Span
from spacy.training import Example
from spacy.pipeline.entity_linker import DEFAULT_NEL_MODEL
import random 
from spacy.util import minibatch, compounding

d = {'countrycode': ['ITA'], 'category_mcc': ['4814'], 'prefix':['Sky'], 'merchant_type':['STANDARD'], 'clean_names':['NOW'], 'stan_names':['SkyTV']}
df = pd.DataFrame(data=d)

def load_entities():

    mcc_dict = defaultdict(list)
    country_dict = defaultdict(list)
    prefix_dict = defaultdict(list)
    merchant_dict = defaultdict(list)
    names = defaultdict(list)


    for index,row in df.iterrows():
        name = row['clean_names'].lower()
        ID = row['stan_names'].lower()
        merchant=row['merchant_type']
        mcc = row['category_mcc']
        country = row['countrycode']
        prefix = row['prefix']
        names[ID].append(name)
        merchant_dict[ID].append(merchant)
        mcc_dict[ID].append(mcc)
        if country!='':
            country_dict[ID].append(country)
        if prefix!='':
            prefix_dict[ID].append(prefix)

    combined_dict = defaultdict(list)

    for d in (merchant_dict, mcc_dict, country_dict, prefix_dict): # you can list as many input dicts as you want here
        for key, value in d.items():
            combined_dict[key].append(set(value))

    for key, value in names.items():
       names[key] = list(set(value))

    # concatenate into a string (for embedding)
    descriptions = defaultdict(list)

    for item in combined_dict.items():
        ID = item[0]
        descriptions[ID] = item[0]+': '+ 'Merchant type is ' + ''.join(item[1][0]) + '. ' + ('Possible mcc category codes are ' if len(item[1][1])>1 else 'Possible mcc category code is ') + ', '.join(item[1][1]) + '. ' + ('Possible country codes are ' if len(item[1][2])>1 else 'Possible country code is ') + ', '.join(item[1][2]) + (''.join(['. Possible prefixes are ',', '.join(item[1][3]), '.'] if len(item[1][3])>1 else ['. Possible prefix is ',', '.join(item[1][3]), '.']) if len(item[1])>3 else '.')
            
            
    return names, descriptions


nlp = spacy.load("it_core_news_md")


def create_kb(vocab):
    
    nlp = spacy.load("it_core_news_md")
    
    kb = KnowledgeBase(vocab, entity_vector_length=300)

    name_dict, desc_dict = load_entities()

    for ID, desc in desc_dict.items():
        desc_doc = nlp(desc)
        desc_enc = desc_doc.vector
        kb.add_entity(entity=ID, entity_vector=desc_enc, freq=300)   
    
    for ID, names in name_dict.items():
        for name in names:
            if name not in kb.get_alias_strings():
                kb.add_alias(alias=name, entities=[ID], probabilities=[1])
                
    return kb




nlp = spacy.load('it_core_news_md')

doc = nlp('0000000128e81d8-i0 now: Merchant type is STANDARD. Mcc category code is 4814. Country code is ITA.')
sentencizer = nlp.add_pipe('sentencizer')

TRAIN_DATA=[]
chars_to_tokens = {}

for token in doc:
    for i in range(token.idx, token.idx + len(token.text)):
        chars_to_tokens[i] = token.i

token_start = chars_to_tokens[19]
token_end = chars_to_tokens[21]
span = Span(doc,token_start,token_end+1)

ID = 'skytv'

annots = {
    'links': {(span.start_char,span.end_char):{ID:1.0}},
    'entities': [(span.start_char, span.end_char, 'ORG')]
}

ex = Example.from_dict(doc, annots)
ex.reference = sentencizer(ex.reference)


TRAIN_DATA.append(ex)



config = {
   "labels_discard": [],
   "n_sents": 0,
   "incl_prior": True,
   "incl_context": True,
   "model": DEFAULT_NEL_MODEL,
   "entity_vector_length": 64,
   "get_candidates": {'@misc': 'spacy.CandidateGenerator.v1'},
}

nlp = spacy.load('it_core_news_md')
entity_linker = nlp.add_pipe("entity_linker", config=config, after='ner').set_kb(create_kb)



pipe_exceptions = ["parser", "tok2vec", "ner", "entity_linker"]
unaffected_pipes = [pipe for pipe in nlp.pipe_names if pipe not in pipe_exceptions]

with nlp.disable_pipes(*unaffected_pipes):  
    optimizer = nlp.resume_training()
    for itn in range(30):  
        random.shuffle(TRAIN_DATA)
        batches = minibatch(TRAIN_DATA, size=compounding(4.0, 32.0, 1.001))  # increasing batch sizes
        losses = {}
        for batch in batches:
            nlp.update(
                batch,   
                drop=0.2,      # prevent overfitting
                losses=losses,
                sgd=optimizer,
            )
            print(itn, "Losses", losses) 

```

I tried a few things (Original state: with `entity_linker`, `ner`, `tok2vec`, `parser` enabled)
1. Disable `entity_linker`: blank error still there
2. Disable `ner`: the model trains, but no loss for `entity_linker`, and 0 for `parser` and `tok2vec`
3. Disable `ner` and `entity_linker`: the model trains, but all losses are 0 (for `parser` and `tok2vec`)
4. Disable `parser` and `tok2vec`: blank error
5. Remove `entities` in the training data: the model trains, but losses for `parser`, `tok2vec` and `entity_linker` are 0
6. Remove `entities` in the training data and disable `parser` and `tok2vec`: the loss dictionary has no value

I would love to use the config system as well but I am very new to command line interface. Would also be great if the team can make more tutorials for newbies. For this project for example, I based it almost entirely on your [notebook](https://github.com/explosion/projects/blob/master/nel-emerson/scripts/notebook_video.ipynb) from last year with a few spacy 3.0 adjustments which obviously did not do a good job given these errors I run into!

**Reply by svlandeg** [reply]
I'll have another detailed look at your code, but in the meantime I'd like to point out that the v3 version of that code is available here: https://github.com/explosion/projects/tree/v3/tutorials/nel_emerson.
If you hadn't found that yet, then yes we should definitely document this better so people find the v3 compatible code!

**Reply by svlandeg** [reply]
Thanks for the code snippet, this was helpful for me to reproduce this.

There's a few different points. First, the `entity_linker` pipeline component is new, and should be [initialized](https://spacy.io/api/entitylinker#initialize):
```
entity_linker = nlp.add_pipe("entity_linker", config=config, after='ner')
entity_linker.set_kb(create_kb)
entity_linker.initialize(lambda: TRAIN_DATA)
```
or
```
entity_linker = nlp.add_pipe("entity_linker", config=config, after='ner')
entity_linker.initialize(lambda: TRAIN_DATA, kb_loader=create_kb)
```

> Disable `ner`: the model trains, but no loss for `entity_linker`, and 0 for `parser` and `tok2vec`

I can't reproduce this part. If the NER is disabled, everything seems to work just fine:
```
pipe_exceptions = ["parser", "tok2vec", "entity_linker"]
```
Gives me
>0 Losses {'tok2vec': 0.0, 'parser': 0.0, 'entity_linker': 0.8664725422859192}
1 Losses {'tok2vec': 0.0, 'parser': 0.0, 'entity_linker': 0.9108008742332458}
...
5 Losses {'tok2vec': 0.0, 'parser': 0.0, 'entity_linker': 0.5233225226402283}
...
10 Losses {'tok2vec': 0.0, 'parser': 0.0, 'entity_linker': 0.33408844470977783}
...
15 Losses {'tok2vec': 0.0, 'parser': 0.0, 'entity_linker': 0.19729453325271606}
...
20 Losses {'tok2vec': 0.0, 'parser': 0.0, 'entity_linker': 0.1027650237083435}
...
29 Losses {'tok2vec': 0.0, 'parser': 0.0, 'entity_linker': 0.0420001745223999}

Do you actually want/need to further train the NER or parser components that you sourced from `it_core_news_md`?

**Reply by wy-yip** [reply]
That's amazing! It works perfectly now. Sorry I was following the code example in the old tutorial and did not know I had to initialize. In the actual project, I trained the NER component and use resume_training() to train the EntityLinker component. Now both of them work and my model is predicting the shop names very well. But I also want to know if this is what you would have done procedure wise? And if I have other relevant information such as location, shop type (e.g. transport, food), will concatenating them into a sentence, which is what I did above, helps linking aliases and entities? Thanks!

**Reply by svlandeg** [reply]
Happy to hear it. It makes a lot of sense to train the NER first, then train the `entity_linker` in a follow-up step. 

For the entity linking, ideally, the training sentences look like actual sentences you'll encounter during prediction time, otherwise the performance in production might not be as good.

**Reply by wy-yip** [reply]
Got it, thanks for all the help!

**Reply by wy-yip** [reply]
am unsure if I should open another discussion for this, but if @svlandeg you are still monitoring this thread, could you tell me if using resume_training() to update the model with a new training dataset is not currently supported? I asked in light of this [post](https://github.com/explosion/spaCy/discussions/6902). I wanted to add more aliases to a few entities in the knowledgebase but am not sure how, since I am no longer running the EntityLinker.initialize() method, and the EntityLinker.update() method is just for example object (which means if the qid is not in the kb, the alias in the example will not be recognised?). Thanks!

**Reply by svlandeg** [reply]
Hi! I don't mind answering here, but the only annoying thing is that github discussions cuts of the thread after X posts, and then you have to click the "read more" button. Anyway.

The entity linker algorithm really has these main steps:
1. create the KB with its aliases and prior probabilities
2. train the entity linker itself to recognize contexts that are similar to descriptions
3. at prediction time: use the KB to generate plausible "candidates" given a certain textual mention
4. use the trained entity linker to decide which of the candidates is the most plausible one.

Now, if I understand your use-case correctly, and you really only want to add aliases to some entities, you can probably get away with changing step 1 (KB creation). As a result, you'll influence step 3 (candidate generation), but the trained entity linker can probably stay as-is.

The tricky bit is, is that you'll need to make sure the updated KB is loaded correctly. You could do this like so:
```
from spacy.ml import load_kb
entity_linker.set_kb(load_kb(kb_dir))
```

If there really is a need for retraining the entity linker itself, or fine-tuning an already trained model, you should be able to call
```
optimizer = nlp.resume_training()
```
but again you'll have to make sure that you have the updated KB set correctly. Because after training, the `nlp` model and its corresponding `entity_linker` pipe serialize all information to one directory. This means that the "old" KB is stored with the model.

**Reply by wy-yip** [reply]
Great, I used get_pipe() and set_kb() and it worked. Thank you!

**Comment by svlandeg** [comment]
Just want to add to this thread, that we've updated [this notebook](https://github.com/explosion/projects/blob/v3/tutorials/nel_emerson/notebooks/notebook_video.ipynb) with the full run-through on how to create and train an entity linker in spaCy v3.

The notebook shows the code if you want to implement/run the training loop in code. If you'd prefer to use the new [config system](https://spacy.io/usage/training#config), you can experiment with the provided [project](https://spacy.io/usage/projects) instead: https://github.com/explosion/projects/blob/v3/tutorials/nel_emerson
